### PR TITLE
CONFLUENCE-28: Introduce entity reference resolver for Confluence reference syntax

### PR DESCRIPTION
**Pull request by ClemensRobbenhaar**
currently defects references of the types:

- mailto
- url
- space
- document
- attachment
- user

It likely does not handle many edge cases but should be better than without any syntax specific resolver.

### CODE REVIEW

**Comment by tmortagne** [comment]
The problem is that this is a breaking change for anyone which is using this syntax in any other context than the Confluence import (for example, in an XWiki page).

The cleanest would be to do the same as what was done for the xhtml syntax parser: ConfluenceXHTMLInputFilterStream is an input filter for that syntax which is used by the XML input filter when parsing Confluence XHTML content and to which are passed some configuration through ConfluenceXHTMLInputProperties. That way, the XML input filter will be able to tell the syntax parser if the reference must be parsed the Confluence way or the standard XWiki rendering way (either with some boolean to choose between the two modes or even by simply passing a custom ResourceReferenceParser, your ConfluenceResourceReferenceParser, which would be implemented on XML filter side).

**Comment by ClemensRobbenhaar** [comment]
Oh, I see, for people who use Confluence wiki syntax to edit pages they need to use the XWiki reference for links, etc. as the Confluence wiki syntax references do not allow things like nested space/pages. My reference resolver indeed makes this impossible.

Reverting PR to draft until this is fixed.

**Comment by ClemensRobbenhaar** [comment]
I have updated the PR.
As the `ConfluenceParser` is a singleton component, I needed to make it clone-able, so the `ConfluenceInputFilterStream` can create a copy which uses the custom confluence reference resolver without influencing the singleton. Somewhat awkward, but that is the best I can come up with.

**Comment by ClemensRobbenhaar** [comment]
After some conversation in chat replaces the ugly `clone` with a separately configured component.

**Comment by ClemensRobbenhaar** [comment]
> Have you tried with a link to a nested page?
> 
> For instance Page 1 is a child of Page A, like this: Space A > Page A > Page 1
> 
> and another page links to Page 1. The produced link should look like "Root.Space A.Page A.Page 1", and not "Root.Space A.Page 1" (when rootSpace is set to Root).

No, I did not do that, because Confluence has no such notation of nested pages. The reference is always of the form `space:page` or just `page`. The nested pages hierarchy in Confluence is done via parent-child relations instead, and not reflected in references to pages.

File: confluence-syntax-confluence/src/test/java/org/xwiki/contrib/confluence/parser/confluence/internal/wikimodel/ConfluenceResourceReferenceParserTest.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        ResourceReference unparseable = new ResourceReference("broken^file^refernce", ResourceType.URL);
+        unparseable.setTyped(false);
+        assertEquals(unparseable,  parser.parse("broken^file^refernce"));
+    }
+
+    @Test
+    public void testDotEscape()
+    {
+        assertEquals(new ResourceReference("space.dot\\.name", ResourceType.DOCUMENT), parser.parse("space:dot.name"));
+        assertEquals( new ResourceReference("dot\\.name", ResourceType.DOCUMENT), parser.parse("dot.name"));
```
[the remark this answers is not in the document]
Comment (ClemensRobbenhaar) [reply confluence-syntax-confluence/src/test/java/org/xwiki/contrib/confluence/parser/confluence/internal/wikimodel/ConfluenceResourceReferenceParserTest.java]: Oh, my, I forgot that checkstyle does not look at test files. Fixed now.

File: confluence-xml/src/main/java/org/xwiki/contrib/confluence/filter/internal/input/ConfluenceInputStreamParser.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+import org.xwiki.contrib.confluence.parser.confluence.internal.ConfluenceParser;
+import org.xwiki.rendering.parser.ResourceReferenceParser;
+
+/**
+ * Parser for Confluence syntax and references.
+ *
+ * Similar to the {@link ConfluenceParser} but also resolves references the Confluence way.
+ *
+ * @version $Id$
+ * @since 9.50.3
```
[the remark this answers is not in the document]
Comment (ClemensRobbenhaar) [reply confluence-xml/src/main/java/org/xwiki/contrib/confluence/filter/internal/input/ConfluenceInputStreamParser.java]: Indeed I forgot that my branch is quite old and there were more releases in the mean time. Fixed as suggested, too.

**Comment by ClemensRobbenhaar** [comment]
I have tested it with an export from an actual instance, and it worked even without the extra `Root.` It seems the importer takes care of the extra step to add the references. (I have to admit that I just trusted the magic here, and did not try to figure out where that happens.)

**Pull request merged**